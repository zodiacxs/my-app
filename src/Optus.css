.loader {
    justify-content: center;
    display: grid;
    margin: 40px;
    text-align: center;
    font-size: 20px;
    font-weight: bold;
}

.OptusTable{
    margin: 100px auto 100px auto;
    width: 80%;
    position:center; 
}

.statsTable {
    margin: 20px;
}

.barTable {
    max-width: 800px;
    position: relative; 
    margin: 20px;
}

.percentage {
    position: absolute; 
    text-align: center;
    width: 100%;
}

.bar {
    cursor: pointer;
    padding: 2px;
    margin: 2px;
    border: 1px solid #333;
    border-radius: 10px;
    width: auto;
    min-width: 200px;
    height: 20px;
    text-align: center;
}

.bar:hover{
    box-shadow: 0 12px 12px 0 rgba(0,0,0,0.24), 0 17px 20px 0 rgba(0,0,0,0.19); 
    background-color: rgb(255, 254, 170); 
}

.filler {
    border-radius: inherit;
    background: rgb(26, 189, 230);
    height: 100%;
    -webkit-transition: 0.5s;
    -moz-transition: 0.5s;
    -o-transition: 0.5s;
    transition: 0.5s;
}

.ButtonRow{
    max-width: 800px;
    margin: 20px;
    text-align: center;
    overflow: auto;
}

button {
    background-color:  transparent;
    border: 3px solid #0099CC;
    border-radius: 10px;
    text-align: center;
    font-size: 30px;
    margin: 2px;
    min-width: 80px;
    padding: 10px; 
    display: inline-blockï¼›
}

button:hover{
    background: rgb(255, 254, 170); 
    box-shadow: 0 12px 16px 0 rgba(0,0,0,0.24), 0 17px 50px 0 rgba(0,0,0,0.19); 
}