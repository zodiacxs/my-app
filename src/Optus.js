import React, { Component } from 'react';
import Loader from 'react-loader-spinner'
import {
    TheLoader,
    OptusTable,
    StatsSection,
    BarSection,
    Bar,
    CoolFiller,
    Percentage,
    ButtonSection,
    NiceButton,
} from "./StyledComp"

const endpoint = 'http://pb-api.herokuapp.com/bars';

export default class Optus extends Component {

    constructor(props){
        super(props);

        this.state = {
            isloading: false,
            buttons: [],
            bars: [],
            limit: 0,
            active: 0,
        };

        this.UpdateProgress = this.UpdateProgress.bind(this);
        this.handleBarClick = this.handleBarClick.bind(this);
    }

    componentDidMount(){
        this.setState({isloading: true});
        fetch (endpoint)
        .then (response => {
            if (response.ok) {
                return response.json();
            }else{
                throw new Error ("Sorry, we have encoutnered techinical difficulty.")
            }
        })
        .then (data => this.setState({
            isloading: false,
            buttons: data.buttons,
            bars: data.bars,
            limit: data.limit,
        }))
        .catch(error => this.setState({error}));
    }

    handleBarClick(index){
        if (this.state.active !== index){this.setState({active:index})}
    }

    UpdateProgress(button){
        const tempBars = this.state.bars;
        tempBars[this.state.active] += button;
        if (tempBars[this.state.active] < 0 ) {tempBars[this.state.active] = 0}
        this.setState({bars: tempBars})
    }

    render(){
        const {isloading, buttons, bars, limit, active} = this.state;

        if(isloading){
            return (
                <TheLoader>
                    <Loader type="Puff" color="#00BFFF" height="200" width="200"/>
                    <p> LOADING </p>
                </TheLoader>
            );
        }

        return(
            <OptusTable>
                <StatsSection>
                    <p>LIMIT: {limit} <br></br> BARS: {bars.toString()}</p>
                </StatsSection>
                <BarSection>
                    {bars.map ((bar, index) =>
                        <ProgressBar
                            key = {index}
                            isActive = {(active+1)/(index+1)}
                            percentage = {Math.floor(bar/limit*100)}
                            onBarClick={this.handleBarClick.bind(null, index)}
                        />
                    )}
                </BarSection>
                <ButtonSection>
                    {buttons.map((button) =>
                        <NiceButton key={button*100} onClick={(e) => this.UpdateProgress(button)}>
                            {button}
                        </NiceButton>
                        
                    )}
                </ButtonSection>
            </OptusTable>
        )
    }
}

const ProgressBar = (props) => {
    let percentage = props.percentage;
    if (props.percentage > 100) {percentage = 100;}

    return(
        <Bar className={props.isActive===1 ? "bar active":"bar"} onClick={props.onBarClick}>
            <Percentage>{percentage}% </Percentage>
            <Filler percentage={percentage}/>
        </Bar>
    )
}

const Filler = (props) => {
    return (
        <CoolFiller 
            className={props.percentage===100? "filler full":"filler"} 
            style={{ width: `${props.percentage}%`}}
        />
    )
}