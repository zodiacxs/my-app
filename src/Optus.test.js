import React from "react";
import Optus from "./Optus"
import Loader from "react-loader-spinner"

describe("Optus", () => {
    
    describe("clickable elements", () => {
        let wrapper;
        const state = {
            isloading: false,
            buttons: [80, 41, -77, -42],
            bars: [11, 44, 53],
            limit: 123,
            active: 0,
        };
        beforeEach(() => {
        wrapper = shallow(<Optus />);
        wrapper.setState(state);
        });
        it('update the bar progress', () => {
            const instance = wrapper.instance();
            for (var i=0; i<4; i++){
                const prevProgress = instance.state.bars[instance.state.active]
                wrapper.find('button').at(i).simulate('click');
                expect (instance.state.bars[instance.state.active]).toEqual(prevProgress+instance.state.buttons[i])
            }
        });
        it('change active bar', () => {
            const instance = wrapper.instance();
            expect(instance.state.active).toEqual(0);
            instance.handleBarClick(2);
            expect(instance.state.active).toEqual(2);
        })
    })

})

