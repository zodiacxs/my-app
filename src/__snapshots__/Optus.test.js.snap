// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="loader"
>
  <Loader
    color="#00BFFF"
    height="100"
    type="Puff"
    width="100"
  />
  <p>
     LOADING 
  </p>
</div>
`;
